tests/h/auth/util_test.py

```diff
@@ -200,6 +202,114 @@ def test_does_not_raise_when_authority_matches(self, pyramid_request, auth_clien
         util.validate_auth_client_authority(auth_client, authority)
 
 
+class TestPrincipalsForAuthClient(object):
+
+    def test_it_sets_auth_client_principal(self, auth_client):
+        principals = util.principals_for_auth_client(auth_client)
+
+        assert "auth_client:{authority}".format(authority=auth_client.authority) in principals
+
+    def test_it_sets_authority_principal(self, auth_client):
+        principals = util.principals_for_auth_client(auth_client)
+
+        assert "authority:{authority}".format(authority=auth_client.authority) in principals
+
+    def test_it_returns_principals_as_list(self, auth_client):
+        principals = util.principals_for_auth_client(auth_client)
+
+        assert isinstance(principals, list)
+
+
+class TestCheckAuthClient(object):
```

Comment:
These tests were quite challenging to write because if you look at [the source for this function](https://github.com/hypothesis/h/pull/5192/files#diff-f54a90f529b1f3f68d72d8c0743466acR146) you'll see that there are a goodly handful of conditions that will cause it to return `None`, meaning the challenge is to mock in such a way that the individual `None` conditions are testable. Not sure if I hit the mark here, but I tried.

Reply:
An alternative approach might have been to not mock sqlalchemy or `models.AuthClient`. That would simplify things somewhat although you'd still need to set up a missing `AuthClient` for one test, an `AuthClient` with no `secret` for another test, etc (I don't think there's any avoiding that, though).

I don't think `h.auth` tests should be accessing the real db, but that's because `h.auth` code shouldn't be using sqlalchemy, it should be calling simpler interfaces provided by our models or services. Given that `h.auth` code _is_ calling sqlalchemy directly I think you can argue either for mocking or for not mocking the db in the tests.